crates/compiler/gen_wasm/src/storage.rs

```diff
@@ -721,7 +721,11 @@ impl<'a> Storage<'a> {
                 debug_assert!(to_size == from_size);
                 self.load_symbols(code_builder, &[from_symbol]);
                 code_builder.set_local(*to_local_id);
-                self.symbol_storage_map.insert(from_symbol, to.clone());
+
+                // Do not repoint the stack memory to the local variable, because it is
+                // possible that other code paths will use the stack memory before this copy.
+                // For example, if the copy happens behind a branch, the local variable may not be
+                // assigned the stack content.
```

Comment:
OK I think there might be some misunderstanding behind this.
Most likely from the highly-overloaded word "stack".

`VirtualMachineStack` means the stack machine, which is best thought of as Wasm's target-independent abstraction over CPU registers. Importantly, this does not involve _memory_ at all.

There are other branches in this `match` that deal with `StackMemory`. This is the usual meaning of "stack" - an area of memory where we allocate space for local variables in procedures. In `gen_wasm` there are lots of places where I consciously say "stack memory" rather than just "stack".

Your comment is in a branch of the `match` that deals with `VirtualMachineStack`, but the text of the comment specifically talks about `StackMemory`. So there is something wrong somewhere. I'm not quite sure what.

This change doesn't look right to me and I suspect it might break lots of tests.

Reply:
`test-gen-wasm` passes for me with this change, but I understand now - this is the current state of the stack machine. Thanks for the explanation.

So, the motivating code here is an IR like

```
joinpoint `#UserApp.9` `#UserApp.f`:
    let `#UserApp.7` : {} = Struct {};
    let `#UserApp.6` : I64 = CallByName `#UserApp.f1` `#UserApp.7` `#UserApp.f`;
    ret `#UserApp.6`;
in
let `#UserApp.11` : Int1 = true;
if `#UserApp.11` then <no branch info>
    jump `#UserApp.9` `#UserApp.x`;
else <no branch info>
    jump `#UserApp.9` `#UserApp.x`;
```

where the two jumps need to store the value of `x` into `f`. Currently, when `x` lives on the stack machine, and `f` is a local, we will repoint `x` to `f` - that happens in the first branch. But, if the first branch is not reached and instead we fall through to the default branch, the wasm codegen has already repointed `x` to `f`'s memory slot, and we will now assign to `f` its own value, which is incorrect. As far as I understand, we need to avoid doing this re-pointing.

So I guess this doesn't work in general because we might enter multiple branches, and once we call `local.set`, the set variable is popped from the active stack - is that correct?

Reply:
Brain and I paired on this. The change was correct, but the explanation in the original comment was incorrect. It's been resolved. Thanks again, Brian.

Reply:
OK after a long discussion, it seems the code change was correct but the comment was wrong!
The old code for copying from one symbol to another was legitimately broken in this case.
It was written based on the idea "now that we've copied the value from the VM stack to a local, we can just use the local instead from now on". But that's wrong for branching code. If we change the storage while generating one branch, it breaks all other branches that are generated later.